odoo/tools/monkeypatch.py

```diff
@@ -0,0 +1,266 @@
+    openssl_version = ssl.OPENSSL_VERSION_INFO[:3]
+
+    lib = OpenSSL._util.lib
+    if pyopenssl_version >= (20, 0, 0) and openssl_version < (1, 1, 0):
+        lib.X509_getm_notBefore = lib.X509_get_notBefore
+        lib.X509_getm_notAfter = lib.X509_get_notAfter
+    elif pyopenssl_version >= (19, 0, 0) and openssl_version >= (1, 1, 0):
+        lib.X509_get_notBefore = lib.X509_getm_notBefore
+        lib.X509_get_notAfter = lib.X509_getm_notAfter
+
+    return OpenSSL
+
+def monkeypatch_PyPDF2():
+    """
+        PyPDF2 fixes:
+        1) Fix invalid numbers like 0.000000000000-5684342
+           Because some PDF generators are building PDF streams with invalid numbers.
+
+        Ref: https://docs.python.org/3/library/zlib.html#zlib.decompressobj
+        2) ensure that zlib does not throw error -5 when decompressing
+           because some pdf won't fit into allocated memory.
+    """
+    import zlib
+    import PyPDF2
+
+    orig_FloatObject___new__ = PyPDF2.generic.FloatObject.__new__
+
+    def FloatObject___new__(cls, value="0", context=None):
+        # Fix invalid numbers like 0.000000000000-5684342
+        # Because some PDF generators are building PDF streams with invalid numbers
+        if isinstance(value, bytes) and value[0] != b'-' and b'-' in value:
+            value = b'-' + b''.join(value.split(b'-', 1))
+        elif isinstance(value, str) and value[0] != '-' and '-' in value:
+            value = '-' + ''.join(value.split('-', 1))
+        return orig_FloatObject___new__(cls, value, context)
+
+    def _decompress(data):
+        zobj = zlib.decompressobj()
+        return zobj.decompress(data)
+
+    PyPDF2.generic.FloatObject.__new__ = FloatObject___new__
+    PyPDF2.filters.decompress = _decompress
+
+    return PyPDF2
+
+def monkeypatch_xlsx():
+    """
+        xlrd.xlsx supports defusedxml, defusedxml's etree interface is broken
+        (missing ElementTree and thus ElementTree.iter) which causes a fallback to
+        Element.getiterator(), triggering a warning before 3.9 and an error from 3.9.
+
+        We have defusedxml installed because zeep has a hard dep on defused and
+        doesn't want to drop it (mvantellingen/python-zeep#1014).
+
+        Ignore the check and set the relevant flags directly using lxml as we have a
+        hard dependency on it.
+    """
+    from xlrd import xlsx
+    from lxml import etree
+
+    xlsx.ET = etree
+    xlsx.ET_has_iterparse = True
+    xlsx.Element_has_iter = True
+
+    return xlsx
+
+def monkeypatch_xlsxwriter():
+    """
+        Xlsxwriter fix, add some sanitization to respect the excel sheet name restrictions
+        as the sheet name is often translatable, can not control the input.
+    """
+    import re
+    import xlsxwriter
+
+    class PatchedXlsxWorkbook(xlsxwriter.Workbook):
+        # TODO when xlsxwriter bump to 0.9.8, add worksheet_class=None parameter instead of kw
```

Comment:
Looks like we bumped this requirement to 1.1.2 in 14.0, either apply the todo or remove it.